### PR TITLE
Fix recreated objects after parked operations

### LINKED ISSUE #19580: Observable#iterator may recreate underlying ringbuffer
**Describe the bug**
Observable#iterator may recreate underlying ringbuffer

**Expected behavior**
ringbuffer should the reliably destroyed.

**To Reproduce**

```java
    @Test
    public void simple() throws ExecutionException, InterruptedException {
        Config config = smallInstanceConfig();
        config.getJetConfig().setEnabled(true);

        HazelcastInstance hz = createHazelcastInstance(config);
        JetService jet = hz.getJet();

        Ringbuffer<Object> ringbuffer = hz.getRingbuffer("__jet.observables.my-observable");

        Observable<Object> obs = jet.getObservable("my-observable");
        ringbuffer.add(42);
        ringbuffer.addAsync(DoneItem.DONE_ITEM, OverflowPolicy.OVERWRITE);

        Iterator<Object> it = obs.iterator();
        it.hasNext();
        it.next();
        it.hasNext();

        obs.destroy();

        HazelcastInstanceProxy proxy = (HazelcastInstanceProxy) hz;
        RingbufferService ringbufferService =
                ((HazelcastInstanceImpl) proxy.getOriginal()).node.getNodeEngine().getService(RingbufferService.SERVICE_NAME);

        ConcurrentMap<Integer, Map<ObjectNamespace, RingbufferContainer>> allContainers =
                ringbufferService.getContainers();

        Map<ObjectNamespace, RingbufferContainer> containers = allContainers.values().iterator().next();
        assertThat(containers).hasSize(0);
    }
```

### PR DESCRIPTION
**Pull request by ramizdundar**
## Introduction

I will explain the problem and offer 3 different solutions. Later I will implement accepted solution.

## Description

During lifetime of the `Observer`, `RingbufferListener` creates many `ReadManyOperation` in a recursive fashion `RingbufferListener#next()`. So we can expect to see a `ReadManyOperation` in the member if there is a data to be read. However if there isn't data to be read then this operation is parked until new data arrives. This is good because otherwise there will be a busy loop.

When `Observable#destroy()` is called, eventually `ProxyServiceImpl#destroyLocalDistributedObject` is called. In this method `service.destroyDistributedObject(name)` called before `nodeEngine.getOperationParker().cancelParkedOperations(serviceName, name, cause)` and this is the problem.

Cancelling parked operations will eventually call `WaitSet#cancelAll` which calls `getWaitKey()` for each blocking operation in the queue. The problem is `ReadManyOperation` is implemented in a way that if underlying ringbuffer isn't found, then a new one will be created. You can see this behaviour in the `AbstractRingBuffer#getRingBufferContainer` method.

## Possible Solutions

Now there are 3 solutions I can think:

1. Call `service.destroyDistributedObject(name)` called after `nodeEngine.getOperationParker().cancelParkedOperations(serviceName, name, cause)`. This is clean but in a place where it could affect too much. Also there may be other places we need to change this order.
2. Re-clean distributed objects after destroy finishes. Since we know that thre will be no longer parked operations this time it should work. This could be done for 3 different scopes: Observables, RingBuffers, all distributed obejcts.
3. Make sure that when `Observable#destroy` is called, there are no parked operations. This solves this problem only for this issue but also the safest.

We can also change underlying behaviour of ReadManyOperation but this could(?) break compatibility.

## Short Description & Implemented Solution

After calling `destroy()` for `Observable` sometimes underlying ringbuffer container isn't destroyed. This is because `ReadManyOperation` may create new containers if the old one isn't found.

Removed parts related to ringbuffer creation in `ReadManyOperation`. Here are the two places effected:

1. `getWaitKey()`: Namespaces are now created for each method call.
2. `shouldWait()`: If the container isn't initailized, `minSize > 0` decides if we are going to wait or not.

`ObservableRingbufferContainerLeakTest` is also added as nightly test.

## Extra

Fixes #19580

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Add `Add to Release Notes` label if changes should be mentioned in release notes or `Not Release Notes content` if changes are not relevant for release notes
- [x] Request reviewers if possible

### CODE REVIEW

**Comment by ramizdundar** [comment]
Note that although we encountered this issue in `Observable`, this issue could be encountered from anywhere that uses `RingBuffer` and `ReadManyOperation`. Also note that if there are other `Operation`s like `ReadManyOperation` which creates the object if it can't find it, then we can see this issue with also that distributed object.

**Comment by vbekiaris** [comment]
Thanks for the analysis @ramizdundar!

Other operations construct their wait keys within `getWaitKey` method implementation (eg for `IMap#put` operation see https://github.com/hazelcast/hazelcast/blob/40d555dc1110da9f63fd66f504c83d4382d0384b/hazelcast/src/main/java/com/hazelcast/map/impl/operation/LockAwareOperation.java#L45-L48). Why do we actually need the ringbuffer container in the `RaadManyOperation` impl?
Also, the same might happen in [`shouldWait`](https://github.com/hazelcast/hazelcast/blob/40d555dc1110da9f63fd66f504c83d4382d0384b/hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/operations/ReadManyOperation.java#L65). Maybe we need a `getRingBufferContainerOrNull()` style method to avoid creating an already destroyed container?

**Comment by ramizdundar** [comment]
### What is the real problem?
Please look at the following code:
```
            // create instance
            Config config = smallInstanceConfig();
            hz = Hazelcast.newHazelcastInstance(config);
            HazelcastInstanceProxy proxy = (HazelcastInstanceProxy) hz;
            RingbufferService ringbufferService = proxy.getOriginal().node.getNodeEngine().getService(RingbufferService.SERVICE_NAME);
            ConcurrentMap<Integer, Map<ObjectNamespace, RingbufferContainer>> allContainers = ringbufferService.getContainers();

            // initialize containers and add one element
            Ringbuffer<Object> ringbuffer = hz.getRingbuffer("any-ringbuffer");
            ringbuffer.add(41);

            // get container list
            Map<ObjectNamespace, RingbufferContainer> containers = allContainers.values().iterator().next();
            
            // test if add creates another ringbuffer
            System.out.println("===" + containers.size());
            ringbuffer.destroy();
            System.out.println("===" + containers.size());
            ringbuffer.add(42);
            System.out.println("===" + containers.size());
```

The observed output is: 
```
===1
===0
===1
```
which means that add method is creating another ringbuffer. This is because most of ringbuffer operations call `AbstractRingBufferOperation#getRingBufferContainer()` which calls `RingbufferService#getOrCreateContainer()`. This also means that most operations after `destroy()` will create the ringbuffer again. In our case it was the parked `ReadManyOperation` was the one creating it.

### `getOrCreateContainer()`
Now this is because of #10901. I understand the problem there and why we need `getOrCreateContainer()`.  As for do we need only for `EventJournal`s or for all of the ringbuffer operations, I'm not sure but maybe @mmedenjak can shed light if he remembers this PR. 

However this causes the problem mentioned before. I also take a look at the other distributed objects. And found this issue: #8546. Now it's mentioned there that: 
> All operations should follow this rule: "If there is a state then always set it eagerly. Do not use if-null-then-set idiom for operation state. "

Our case is one counter example of this. However I also do realize that solving #10901 without this could be too messy.

> Why do we actually need the ringbuffer container in the RaadManyOperation impl?

I believe this is because we need something from the storage to identify what is waiting. In IMap case we take it from the recordStore but queue also takes it from the underlyin container.

### What could we do?

1. Accept and document that ringbuffer operations will create a new container after destroy. And enhance destroy for parked operations.
2. Fail fast in most of the ringbuffer operations if destroy is called before. I don't know how this would interact with race conditions or can we determine if destroy() is called before or not.
3. Do not use` getOrCreateContainer()`. But this is hard because we don't know what would it cause in the future.
4. Any other suggestions?

cc @mmedenjak @jerrinot because two of PR mentioned here, we could use your expertise. Feel free to ignore if you don't have free time.

**Comment by vbekiaris** [comment]
In general `destroy()` may be called by any thread (eg from an event thread as a result of a distributed object destroyed event received from a remote member), while ringbuffer ops are invoked on a partition thread. Coordinating between object creation/destruction & partition operations can be quite complex and the fact that our distributed objects (not just ringbuffer) do not have a well-defined lifecycle is a known issue (eg see #6832 , #9748). Determining a general solution for object lifecycle is definitely outside the scope of the issue at hand.

Regarding the specific issue, let me rephrase my question: Why do we actually need the ringbuffer container in the `ReadManyOperation#getWaitKey()` implementation? 

We don't really need it; we could instead `return new DistributedObjectNamespace(SERVICE_NAME, name);` (see `RingbufferService#getRingbufferNamespace`). The fact that we use the ringbuffer container in `getWaitKey()` is just an optimization to avoid creating litter. If the ringbuffer container does not exist, then we can just create & return a new `ObjectNamespace`. Avoiding recreating the ringbuffer container in `getWaitKey` should be enough to allow parked operations to be cancelled without resurrecting a destroyed container.

**Review by frant-hartm — APPROVED** [review]

File: hazelcast/src/test/java/com/hazelcast/jet/core/ObservableRingbufferContainerLeakTest.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(NightlyTest.class)
+public class ObservableRingbufferContainerLeakTest extends JetTestSupport {
+
+    /**
+     * Regression test for #19580
+     */
+    @Test
+    @Repeat(100)
```
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/jet/core/ObservableRingbufferContainerLeakTest.java:50]: leftover repeat?

**Review by vbekiaris — APPROVED** [review]
Thanks for the fix! Should we also apply it to `ReadOneOperation`?

File: hazelcast/src/test/java/com/hazelcast/jet/core/ObservableRingbufferContainerLeakTest.java (hunk 1 above)
Comment (ramizdundar) [reply hazelcast/src/test/java/com/hazelcast/jet/core/ObservableRingbufferContainerLeakTest.java:50]: No this is here because the test almost never fails on the first try because of the race conditions. Even with the repeats the test takes around 5 secs.

Comment (frant-hartm) [reply hazelcast/src/test/java/com/hazelcast/jet/core/ObservableRingbufferContainerLeakTest.java:50]: No, the test doesn't reproduce it 100 %, so I added the repeat. 100 was quite reliable on my machine.

It is marked `NightlyTest` not to interfere with the PR builder time (although it's just few seconds).

**Comment by vbekiaris** [comment]
what about this:

> Should we also apply it to `ReadOneOperation`?

**Comment by ramizdundar** [comment]
It's not necessary for Observable. I will create a seperate PR for that if necessary since jet soak tests may depend on this.

**Comment by mmedenjak** [comment]
Manager is late to the party again :) I just read the comments and Vassilis hit the nail on the head - the proxy lifecycle is messy, and I was just trying to fix one of Danny's test failures, not really trying to address the bigger issue (the proxy lifecycle) and I didn't really have any deeper meaning to the PR other than fixing concurrent destroy and read operation. Unless we decide to tackle the proxy lifecycle issue on a global scale, I think any best-effort (not perfect) solution around ringbuffers is good enough, and seems you were able to find it :)

**Comment by ramizdundar** [comment]
Can someone with the knowledge elaborate why PR builder was triggered?

**Comment by frant-hartm** [comment]
Seems like PR title change triggers PR build.

**Comment by pveentjer** [comment]
@vbekiaris @mmedenjak is this fix going to be backported? This issue seems to be related:
https://github.com/hazelcast/hazelcast/issues/20055

**Comment by mmedenjak** [comment]
Sure, if @ramizdundar considers it safe to backport, and if it fixes the issue, let's backport.

**Comment by ramizdundar** [comment]
I think it should be safe but I don't know if it's simple. https://github.com/hazelcast/hazelcast/pull/19788 and https://github.com/hazelcast/hazelcast/pull/19839 should be backported as well they have more general approach.

1. https://github.com/hazelcast/hazelcast/pull/19630 investigation and fixing one specific problem.
2. https://github.com/hazelcast/hazelcast/pull/19788 general solution to the problem found in https://github.com/hazelcast/hazelcast/pull/19630.
3. https://github.com/hazelcast/hazelcast/pull/19839 parts missing in https://github.com/hazelcast/hazelcast/pull/19788.

cc @frant-hartm
